### ISSUE TITLE
trtexec dynamic batch size

### ISSUE THREAD

**Issue by zirui** [NVIDIA/TensorRT#976]
## Description
 I tried to convert my  onnx model to  tensorRT model with trtexec , and i want the batch size to be dynamic,
 but failed with two problems:
1. trtrexec with `maxBatch` param failed
2.  tensorRT model was converted successfully after specify shape params(minShapes/maxShapes/optShapes),  but error occurs when load model back with trtexec

## Environment

**TensorRT Version**:  7.2.1
**GPU Type**:  Telsa P40
**Nvidia Driver Version**:  418.40.04 
**CUDA Version**:  cuda11.1
**CUDNN Version**:  8.04
**Operating System + Version**: 18.04.5 LTS 
**Python Version (if applicable)**: 3.6.6
**Baremetal or Container (if container which image + tag)**: nvcr.io/nvidia/tensorrt:20.11-py3

### 1. Convert onnx model with `maxBatch`
```
trtexec --onnx=bert.onnx --maxBatch 16 --saveEngine=bert.trt
```
error logs returned by trtexec 
```
[12/20/2020-07:31:45] [E] stoi
```

### 2. 
(1) Convert to tensorRT model by specify Shape params
```
trtexec --onnx=bert.onnx \
        --explicitBatch \
        --saveEngine=bert_model.trt \
        --minShapes=input_ids:1x128,attention_mask:1x128,token_type_ids:1x128 \
        --optShapes=input_ids:8x128,attention_mask:8x128,token_type_ids:8x128 \
        --maxShapes=input_ids:16x128,attention_mask:16x128,token_type_ids:16x128 \
```
(2) Load model with trtexec
```
trtexec --loadEngine=bert_model.trt
```
```
[12/20/2020-07:19:07] [E] Error opening engine file:bert_model.trt
[12/20/2020-07:19:07] [E] Engine creation failed
[12/20/2020-07:19:07] [E] Engine set up failed
&&&& FAILED TensorRT.trtexec # trtexec --loadEngine=bert_model.trt
```

**Comment by nvpohanh** [comment]
(1) `maxBatch` cannot be used with ONNX, which only supports explicit batch dimension. So we'll focus on (2).

(2) When you load the engine, you need to pass in the actual shape you would like to run the inference with by adding `--shapes=input_ids:<N>x128,attention_mask:<N>x128,token_type_ids:<N>x128`, where `<N>` is the actual batch size for inference. Could you try this and report if it works? Thanks

**Comment by zirui** [comment]
thanks @nvpohanh 
i added '--shapes' as you said,   and trtexec run inference  successfully.
```
trtexec --loadModel=bert_model.trt \
        --shapes=input_ids:32x128,attention_mask:32x128,token_type_ids:32x128
```

but the  inference time is more than 50x than trt model with fixed `batch size 1`(converted without specify minShape/Maxshape parms), 
so now i have another two questions:
1. does the inference time returned by trtexec is the `total time of inference <N>  samples` rather than the time of inference  every `single sample`?
2. is it normal that dynamic batch model(N >1) is slower than model with fixed batch size of 1 when inference single sample

GPU latency:
|  model |batch size   | inference time (ms)  |
|---|---|---|
|default model(batch size = 1)  | 1  |   7|
|dynamic batch model   |  1 |20   |
| dynamic batch model | 32  | 380  |


message  of   dynamic batch model when run inference with shapes params
```
[12/21/2020-13:12:28] [I] Average on 10 runs - GPU latency: 380.035 ms - Host latency: 380.084 ms (end to end 759.362 ms, enqueue 236.376 ms)
[12/21/2020-13:12:28] [I] Host Latency
[12/21/2020-13:12:28] [I] min: 379.533 ms (end to end 754.411 ms)
[12/21/2020-13:12:28] [I] max: 380.833 ms (end to end 761.23 ms)
[12/21/2020-13:12:28] [I] mean: 380.084 ms (end to end 759.362 ms)
[12/21/2020-13:12:28] [I] median: 379.969 ms (end to end 759.445 ms)
[12/21/2020-13:12:28] [I] percentile: 380.833 ms at 99% (end to end 761.23 ms at 99%)
[12/21/2020-13:12:28] [I] throughput: 0 qps
[12/21/2020-13:12:28] [I] walltime: 4.17524 s
[12/21/2020-13:12:28] [I] Enqueue Time
[12/21/2020-13:12:28] [I] min: 231.493 ms
[12/21/2020-13:12:28] [I] max: 237.601 ms
[12/21/2020-13:12:28] [I] median: 236.726 ms
[12/21/2020-13:12:28] [I] GPU Compute
[12/21/2020-13:12:28] [I] min: 379.502 ms
[12/21/2020-13:12:28] [I] max: 380.762 ms
[12/21/2020-13:12:28] [I] mean: 380.035 ms
[12/21/2020-13:12:28] [I] median: 379.915 ms
[12/21/2020-13:12:28] [I] percentile: 380.762 ms at 99%
[12/21/2020-13:12:28] [I] total compute time: 3.80035 s
```

**Comment by nvpohanh** [comment]
trtexec returns the runtime per inference, where an "inference" is a query of batch_size=N which you specified. 

> is it normal that dynamic batch model(N >1) is slower than model with fixed batch size of 1 when inference single sample

It is possible, although in this case I would be surprised by this difference. Could you share what commands you use to build the "default model" and the "dynamic batch (batch=1) model"? The default should be identical (or equivalent to) using `--minShapes=input_ids:1x128,attention_mask:1x128,token_type_ids:1x128 --optShapes=input_ids:1x128,attention_mask:1x128,token_type_ids:1x128 --maxShapes=input_ids:1x128,attention_mask:1x128,token_type_ids:1x128` flags.

**Comment by zirui** [comment]
i used below commands to build and test models:
**build model**
default model:
```
trtexec --onnx=bert_model.onnx --saveEngine=model_default.trt
```

dynamic batch model:
```
trtexec --onnx=bert_model.onnx \
        --explicitBatch \
        --saveEngine=batch_model.trt \
        --minShapes=input_ids:1x128,attention_mask:1x128,token_type_ids:1x128 \
        --optShapes=input_ids:8x128,attention_mask:8x128,token_type_ids:8x128 \
        --maxShapes=input_ids:64x128,attention_mask:64x128,token_type_ids:64x128 \
```
-----
**inference test**
default model:
```
trtexec --loadEngine=model_default.trt
```

batch model(inference with batch size of 1  and 32)
```
trtexec --loadEngine=batch_model.trt --shapes=input_ids:1x128,attention_mask:1x128,token_type_ids:1x128
trtexec --loadEngine=batch_model.trt --shapes=input_ids:32x128,attention_mask:32x128,token_type_ids:32x128
```

**Comment by nvpohanh** [comment]
I see. Since "optShapes" is batch_size=8, TRT will use batch_size=8 to choose the fastest tactics, which may not be optimal for batch_size=1 and batch_size=32. Could you try to change optShapes from 8 to 1 or 32 and see if that helps improve the inference time?

**Comment by zirui** [comment]
By  changed optShapes from 8 to 1 and 32, I build another two models , 
and got the preformance report below:

|  model |build-maxShapes|build-optShapes| inference-batch_size  | median GPU latency (ms)  |
|---|---|---|---|---|
|default model |default(1)  | default(1)  |  1|7|
|dynamic batch model| 64|8|    1 |20   |
| dynamic batch mode|64|8 | 32  | 380  |
| dynamic batch mode|64|1| 1  |  20.1 |
| dynamic batch mode|64|1| 32  | 379.4  |
| dynamic batch mode|64|32 | 1  | 20  |
| dynamic batch mode|64|32| 32| 380.3  |

it seems like that the `optShapes` doesn't make any difference

**Comment by nvpohanh** [comment]
I see. That probably means there are some faster tactics taking advantage of static shapes but won't work with dynamic shapes. Could you try setting max=opt=min=inference=32 and see if that helps the latency for batch_size=32 case?

**Comment by zirui** [comment]
i tried ` max=opt=min=inference=32`, then the gpu latency increase to 472ms!
and i noticed that there was warning message in the process of building model with trtexec
```
[I] [TRT] Some tactics do not have sufficient workspace memory to run. Increasing workspace size may increase performanc
```
but trtexec always return `[E] stoi` when add the `workspace` param
```
 trtexec --onnx=bert.onnx --saveEngine=bert.trt --workspace 4096  --shapes=input_ids:32x128,attention_mask:32x128,token_type_ids:32x128
```

**Comment by hexiaoyi95** [comment]
> i tried ` max=opt=min=inference=32`, then the gpu latency increase to 472ms!
> and i noticed that there was warning message in the process of building model with trtexec
> 
> ```
> [I] [TRT] Some tactics do not have sufficient workspace memory to run. Increasing workspace size may increase performanc
> ```
> 
> but trtexec always return `[E] stoi` when add the `workspace` param. eg:
> 
> ```
>  trtexec --onnx=bert.onnx --saveEngine=bert.trt --workspace 4096  --shapes=input_ids:32x128,attention_mask:32x128,token_type_ids:32x128
> ```

Same error occured when setting the workspace.

**Comment by nvpohanh** [comment]
@zirui Could you try `--workspace=4096` instead of `--workspace 4096`? Thanks

**Comment by zirui** [comment]
> @zirui Could you try `--workspace=4096` instead of `--workspace 4096`? Thanks

it works! 
i think we  need more detail error logs, the message `[E] stoi` is too confusing..

And when  test with the new model built with `--workspace=4096`,
the inference time reduce to 380ms, which seems better than 472 , but no difference with opt= 1/8/32

**Comment by nvpohanh** [comment]
I see. I think now the mystery is: does `max=opt=min=inference=1` give you 7ms or 20ms?

**Comment by zirui** [comment]
i rebuild model with differenct `min/opt/max shapes(include max=opt=min=1)`, and update the preformance report below:

min/opt/shapes:
```
1-1-1
1-1-64
1-8-64
1-32-64
32-32-32
```

peformance test

|  model |build/min-opt-max|inference/batch_size  | mean GPU latency (ms)  |
|---|---|---|---|
|default model |default | 1|7.01869|
| dynamic batch mode|1-1-1| 1|19.7453||
| dynamic batch mode|1-1-1| 32|19.745|
| dynamic batch mode|1-8-64| 1 | 20.0619  |
| dynamic batch mode|1-8-64|32  | 379.261  |
| dynamic batch mode|1-1-64| 1 | 19.9791|
| dynamic batch mode|1-1-64|32  | 381.134  |
| dynamic batch mode|1-32-64| 1 | 20.0532  |
| dynamic batch mode|1-32-64|32  | 378.898  |
| dynamic batch mode|32-32-32| 1|378.634  |
| dynamic batch mode|32-32-32| 32|378.695 |
| dynamic batch mode|8-32-32| 1|378.634  |
| dynamic batch mode|8-32-32| 32|380.618 |


so, from the table, we can see that:
model with `max=opt=min=inference=1 ` cost around 20 ms, while the default model cost 7ms

**Comment by nvpohanh** [comment]
@zirui Would it be possible that you can provide the onnx file so that we can debug this gap? Thanks

**Comment by zirui** [comment]
> @zirui Would it be possible that you can provide the onnx file so that we can debug this gap? Thanks

I am unable to  provide the model file due to  the restriction of my private network environment,
but i will take time to train a new model in public network environment  to reproduce the problem this week

**Comment by nvpohanh** [comment]
Thanks. I think the main focus would be: why is 1-1-1 different from the default model?

**Comment by zirui** [comment]
hi,  @nvpohanh 
I have rebuilt the onnx model and uploaded the model to the [google drive](https://drive.google.com/file/d/1arkksDCOHyf1Yq1MpP46u1CmLzBwXmp6/view?usp=sharing)

**Comment by nvpohanh** [comment]
@zirui I see that your onnx file input shapes have two dynamic axes: `[batch, sequence]`, so when you don't pass in any shapes, trtexec actually sets the dynamic axes to 1 and throws a warning in the logs. That means, your 7ms measurement was in fact measuring the performance when the input shape is [1, 1].

![2021-01-25 13_17_06-C__Users_phuan_Desktop_work_ocnli onnx - Netron](https://user-images.githubusercontent.com/53919306/105664479-baea5680-5f0f-11eb-8627-678057b77526.png)

**Comment by zirui** [comment]
I  saw the warning logs when build `default` tensorrt model:
```
[W] Dynamic dimensions required for input: token_type_ids, but no shapes were provided. Automatically overriding shape to: 1x1
```

but i found that  the model could run inference by specify `[batch, sequence]` to `1 x 128` without any error or warning, 
and the inference time is around 7ms
```
trtexec --loadEngine=default.trt --shapes=input_ids:1x128,attention_mask:1x128,token_type_ids:1x128
```


```
[01/25/2021-09:28:53] [I] === Inference Options ===                                                                                  [0/2587]
[01/25/2021-09:28:53] [I] Batch: Explicit
[01/25/2021-09:28:53] [I] Input inference shape: token_type_ids=1x128
[01/25/2021-09:28:53] [I] Input inference shape: input_ids=1x128
[01/25/2021-09:28:53] [I] Input inference shape: attention_mask=1x128
...
[01/25/2021-09:29:23] [I] GPU Compute
[01/25/2021-09:29:23] [I] min: 7.00513 ms
[01/25/2021-09:29:23] [I] max: 7.17107 ms
[01/25/2021-09:29:23] [I] mean: 7.03121 ms
[01/25/2021-09:29:23] [I] median: 7.02771 ms
[01/25/2021-09:29:23] [I] percentile: 7.08594 ms at 99%

```

so does this mean the default model run inference with fixed `1 x 1`input, and the `shapes` params actually dose not work as expected?

**Comment by nvpohanh** [comment]
This is because the engine has been built with a **fixed** shape of 1x1, so the input shape is no longer dynamic. That also means trtexec won't try to set the input shape again (it's already static, anyway). I do think we should improve trtexec so that it throws errors (or at least warnings) if user-provided shapes are not used.

**Comment by zirui** [comment]
@nvpohanh
thanks for your kindly reply, i'm closing this issue

**Comment by Vampire-Vx** [comment]
> (1) `maxBatch` cannot be used with ONNX, which only supports explicit batch dimension. So we'll focus on (2).
> 
> (2) When you load the engine, you need to pass in the actual shape you would like to run the inference with by adding `--shapes=input_ids:<N>x128,attention_mask:<N>x128,token_type_ids:<N>x128`, where `<N>` is the actual batch size for inference. Could you try this and report if it works? Thanks

Hey, you said that "ONNX, which only supports explicit batch dimension." 
I try the command `trtexec --onnx=bert_model.onnx --saveEngine=model_default.trt --batch=5`. It works and I can loadEngine to run inference for batch<=5. And I am confused is it implicit or explicit?

**Comment by nvpohanh** [comment]
Starting from TRT 7 (I think), ONNX is always explicit batch. Using `--batch` has no effect. You can check the throughput or latency numbers and see that it makes no difference

**Comment by KDr2** [comment]
My input shape is 1x2x2x20, and the output shape is (1,).  I can use `trtexec --shape=input:<N>x2x2x20` to specify a batch size N, but the output is still (1,), how to specify its shape to (N,)? 

Thanks.

**Comment by nvpohanh** [comment]
@KDr2 When you export the ONNX file, could you specify the batch dimension to be a dynamic dimension? If you are using PyTorch, then use the `dynamic_axes` argument in `torch.onnx.export()` function call: https://pytorch.org/tutorials/advanced/super_resolution_with_onnxruntime.html

**Comment by KDr2** [comment]
Hi @nvpohanh, my onnx model is converted from a tensorflow saved mode:
```
    inputs['input_2'] tensor_info:
        dtype: DT_INT32
        shape: (-1, 2, 2, 20)
        name: texts:0
```
Above is the tensor info of the input.

And then I used `python -m tf2onnx.convert --saved-model saved_model/ --output 2.onnx  --opset=11` to convert it to an ONNX. Is there a way to make the batch dimension dynamic in this approach?

Thanks.

**Comment by nvpohanh** [comment]
I am not very familiar with tf2onnx, but it seems that you can try this:

> --inputs, --outputs
> TensorFlow model's input/output names, which can be found with summarize graph tool. Those names typically end with :0, for example --inputs input0:0,input1:0. Inputs and outputs are not needed for models in saved-model format. Some models specify placeholders with unknown ranks and dims which can not be mapped to onnx. In those cases one can add the shape after the input name inside [], for example --inputs X:0[1,28,28,3]. Use -1 to indicate unknown dimensions.

https://github.com/onnx/tensorflow-onnx#--inputs---outputs

**Comment by sunshineflg** [comment]
> loadEngine

excuse me, are you compare inf time with torch model and trt model?

**Comment by sunshineflg** [comment]
> I saw the warning logs when build `default` tensorrt model:
> 
> ```
> [W] Dynamic dimensions required for input: token_type_ids, but no shapes were provided. Automatically overriding shape to: 1x1
> ```
> 
> but i found that the model could run inference by specify `[batch, sequence]` to `1 x 128` without any error or warning, and the inference time is around 7ms
> 
> ```
> trtexec --loadEngine=default.trt --shapes=input_ids:1x128,attention_mask:1x128,token_type_ids:1x128
> ```
> 
> ```
> [01/25/2021-09:28:53] [I] === Inference Options ===                                                                                  [0/2587]
> [01/25/2021-09:28:53] [I] Batch: Explicit
> [01/25/2021-09:28:53] [I] Input inference shape: token_type_ids=1x128
> [01/25/2021-09:28:53] [I] Input inference shape: input_ids=1x128
> [01/25/2021-09:28:53] [I] Input inference shape: attention_mask=1x128
> ...
> [01/25/2021-09:29:23] [I] GPU Compute
> [01/25/2021-09:29:23] [I] min: 7.00513 ms
> [01/25/2021-09:29:23] [I] max: 7.17107 ms
> [01/25/2021-09:29:23] [I] mean: 7.03121 ms
> [01/25/2021-09:29:23] [I] median: 7.02771 ms
> [01/25/2021-09:29:23] [I] percentile: 7.08594 ms at 99%
> ```
> 




> @nvpohanh thanks for your kindly reply, i'm closing this issue

excuse me, are you compare inf time with torch model and trt model?

**Comment by sunshineflg** [comment]
> @nvpohanh thanks for your kindly reply, i'm closing this issue

excuse me, are you compare inf time with torch model and trt model?

**Comment by lix19937** [comment]
> excuse me, are you compare inf time with torch model and trt model?

usually infer time at torch model <  trt model

**Comment by jinwater88** [comment]
@ttyio @nvluxiaoz 

The mnist.onnx and resnet50.onnx models in the TensorRT examples are usually static batch (batch size = 1) models, which means it does not support dynamic batch inference by default?
I am doing stress testing. If I want to use the onnx model inference to occupy the entire GPU, and the static model does not support batch processing, is there any other way? Use the command trtexec
